### ISSUE TITLE
azurerm_role_assignment gets destroyed and recreated during azurerm_kubernetes_cluster update

### ISSUE THREAD

**Issue by masterphenix** [hashicorp/terraform-provider-azurerm#13152]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v1.0.5
on linux_amd64
+ provider registry.terraform.io/hashicorp/azuread v1.6.0
+ provider registry.terraform.io/hashicorp/azurerm v2.73.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.4.1
+ provider registry.terraform.io/hashicorp/null v3.1.0
```

### Affected Resource(s)

* resource `azurerm_role_assignment`
* resource `azurerm_kubernetes_cluster`
* data `azuread_service_principal`

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "= 2.73.0"
    }
    azuread = {
      source  = "hashicorp/azuread"
      version = "= 1.6.0"
    }
  }

  backend "azurerm" {
    subscription_id = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
    tenant_id       = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
    resource_group_name   = "terraform-EUW-sandbox-RG"
    storage_account_name  = "xxxxxxxxxx01"
    container_name        = "tfstate"
    key                   = "terraform.tfstate"
  }
}

provider "azurerm" {
  subscription_id = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
  tenant_id       = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
  features {}
}

data "azurerm_subscription" "current" {

}

provider "azuread" {
  environment = "public"
}

data "azuread_group" "aks-cluster-admin" {
  display_name = "g.kubernetes.sandbox.admin"
}

data "azurerm_virtual_network" "vnet" {
  name                = "sandbox-VNet"
  resource_group_name = "sandbox-RG"
}

data "azurerm_public_ip_prefix" "public-ip-prefix" {
  name                = "sandbox-prefix"
  resource_group_name = "sandbox-RG"
}

locals {
  aks_cluster_name          = "myakscluster"
  aks_cluster_nodepool_name = "myakscluster"
}

resource "azurerm_resource_group" "aks-rg" {
  name     = "${local.aks_cluster_name}-rg"
  location = "westeurope"
}

resource "azurerm_public_ip" "aks-outbound-pub-ip" {
  name                = "${local.aks_cluster_name}-outbound-ippub"
  resource_group_name = azurerm_resource_group.aks-rg.name
  location            = azurerm_resource_group.aks-rg.location
  allocation_method   = "Static"
  sku                 = "Standard"
  public_ip_prefix_id = data.azurerm_public_ip_prefix.public-ip-prefix.id
}

resource "azurerm_subnet" "aks-agent-subnet" {
  name                 = "${local.aks_cluster_name}-agent-subnet"
  address_prefixes     = ["10.102.10.0/23"]
  virtual_network_name = data.azurerm_virtual_network.vnet.name
  resource_group_name  = data.azurerm_virtual_network.vnet.resource_group_name
  service_endpoints    = ["Microsoft.Sql"]
}

resource "azurerm_kubernetes_cluster" "aks-cluster" {
  lifecycle {
    ignore_changes = [default_node_pool.0.node_count]
  }

  name                            = local.aks_cluster_name
  location                        = azurerm_resource_group.aks-rg.location
  resource_group_name             = azurerm_resource_group.aks-rg.name
  dns_prefix                      = local.aks_cluster_name
  kubernetes_version              = "1.20.7"
  api_server_authorized_ip_ranges = []
  
  default_node_pool {
    name                 = "${local.aks_cluster_nodepool_name}01"
    vm_size              = "Standard_D4s_v3"
    availability_zones   = [1]
    orchestrator_version = "1.20.7"
    enable_auto_scaling  = true
    min_count            = "1"
    max_count            = "3"
    node_count           = "1"
    type                 = "VirtualMachineScaleSets"
    max_pods             = "50"
    os_disk_size_gb      = "128"
    vnet_subnet_id       = azurerm_subnet.aks-agent-subnet.id
    node_labels          = {}
  }

  linux_profile {
    admin_username = "sandbox"
    ssh_key {
      key_data = "ssh-rsa xxxxxxx== xxxxxx@xxxxx.xxxx"
    }
  }

  network_profile {
    network_plugin     = "azure"
    network_policy     = "calico"
    dns_service_ip     = "10.200.0.10"
    docker_bridge_cidr = "172.17.0.1/16"
    outbound_type      = "loadBalancer"
    service_cidr       = "10.200.0.0/16"
    load_balancer_sku  = "standard"
    load_balancer_profile {
      outbound_ip_address_ids = [azurerm_public_ip.aks-outbound-pub-ip.id]
    }
  }

  addon_profile {
    oms_agent {
      enabled = false
    }
    kube_dashboard {
      enabled = false
    }
  }

  role_based_access_control {
    enabled = true
    azure_active_directory {
      managed                = true
      client_app_id          = null
      server_app_id          = null
      server_app_secret      = null
      admin_group_object_ids = [data.azuread_group.aks-cluster-admin.id]
    }
  }

  identity {
    type = "SystemAssigned"
  }
}

data "azuread_service_principal" "aks-enterprise-application" {
  display_name = local.aks_cluster_name
  depends_on = [
    azurerm_kubernetes_cluster.aks-cluster,
  ]
}

data "azuread_service_principal" "aks-enterprise-application-managed-id" {
  display_name = "${local.aks_cluster_name}-agentpool"
  depends_on = [
    azurerm_kubernetes_cluster.aks-cluster,
  ]
}

resource "azurerm_role_assignment" "aks-rg-contributor" {
  scope                = "/subscriptions/${data.azurerm_subscription.current.subscription_id}"
  role_definition_name = "Contributor"
  principal_id         = data.azuread_service_principal.aks-enterprise-application.id
}

resource "azurerm_role_assignment" "aks-rg-network-contributor" {
  scope                = "/subscriptions/${data.azurerm_subscription.current.subscription_id}"
  role_definition_name = "Network Contributor"
  principal_id         = data.azuread_service_principal.aks-enterprise-application.id
}

# The user assigned managed identity created and used by the cluster needs acrPull role on the Azure Registries
# to be able to pull Docker images
resource "azurerm_role_assignment" "aks-agentpool-rg-acr-acrpull" {
  scope                = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ACR-RG"
  role_definition_name = "AcrPull"
  principal_id         = data.azuread_service_principal.aks-enterprise-application-managed-id.id
}
```

### Expected Behaviour

azurerm_role_assignment resources should not be destroyed and recreated, as they are not modified.

### Actual Behaviour

On first run, everything is created correctly.

Then, if I change something on the cluster (for example, the default_node_pool max_count), and apply my changes, here is what happens :

```hcl
Terraform will perform the following actions:

  # data.azuread_service_principal.aks-enterprise-application will be read during apply
  # (config refers to values not yet known)
 <= data "azuread_service_principal" "aks-enterprise-application"  {
      ~ app_roles      = [] -> (known after apply)
      ~ application_id = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ id             = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ object_id      = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
        # (1 unchanged attribute hidden)
    }

  # data.azuread_service_principal.aks-enterprise-application-managed-id will be read during apply
  # (config refers to values not yet known)
 <= data "azuread_service_principal" "aks-enterprise-application-managed-id"  {
      ~ app_roles      = [] -> (known after apply)
      ~ application_id = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ id             = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ object_id      = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
        # (1 unchanged attribute hidden)
    }

  # azurerm_kubernetes_cluster.aks-cluster will be updated in-place
  ~ resource "azurerm_kubernetes_cluster" "aks-cluster" {
        id                                  = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourcegroups/myakscluster-rg/providers/Microsoft.ContainerService/managedClusters/myakscluster"
        name                                = "myakscluster"
        tags                                = {}
        # (17 unchanged attributes hidden)



      ~ default_node_pool {
          ~ max_count                    = 3 -> 4
            name                         = "myakscluster01"
            tags                         = {}
            # (19 unchanged attributes hidden)
        }

        # (8 unchanged blocks hidden)
    }

  # azurerm_role_assignment.aks-agentpool-rg-acr-acrpull must be replaced
-/+ resource "azurerm_role_assignment" "aks-agentpool-rg-acr-acrpull" {
      ~ id                               = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ACR-RG/providers/Microsoft.Authorization/roleAssignments/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ name                             = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ principal_id                     = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply) # forces replacement
      ~ principal_type                   = "ServicePrincipal" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      + skip_service_principal_aad_check = (known after apply)
        # (2 unchanged attributes hidden)
    }

  # azurerm_role_assignment.aks-rg-contributor must be replaced
-/+ resource "azurerm_role_assignment" "aks-rg-contributor" {
      ~ id                               = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleAssignments/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ name                             = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ principal_id                     = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply) # forces replacement
      ~ principal_type                   = "ServicePrincipal" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      + skip_service_principal_aad_check = (known after apply)
        # (2 unchanged attributes hidden)
    }

  # azurerm_role_assignment.aks-rg-network-contributor must be replaced
-/+ resource "azurerm_role_assignment" "aks-rg-network-contributor" {
      ~ id                               = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleAssignments/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ name                             = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      ~ principal_id                     = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply) # forces replacement
      ~ principal_type                   = "ServicePrincipal" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> (known after apply)
      + skip_service_principal_aad_check = (known after apply)
        # (2 unchanged attributes hidden)
    }

Plan: 3 to add, 1 to change, 3 to destroy.
```

All role assignements are destroyed, then the cluster is updated in place, and then the role assignements are recreated.

But the principal_id does not change, so, to me, role assignements should remain unchanged.

### Steps to Reproduce

1. `terraform apply` a first time, to create the resources
2. Change max_count in the default_nodepool from 3 to 4 (for example)
3. `terraform apply` again

### References

* #0000

**Comment by myc2h6o** [comment]
I've been trying this myself. After reading https://www.terraform.io/language/data-sources#data-resource-dependencies and https://github.com/hashicorp/terraform/issues/30083, I believe the issue is because of the dependency behavior on Terraform data source. According to the document, if a data source directly depends on another managed resource, it will be recalculated during the apply phase because it doesn't know for sure whether the upstream is gonna change or not. The workaround would be use a local value to create an independent dependency. e.g.:

After changing:
``` hcl
data "azuread_service_principal" "aks-enterprise-application" {
  display_name = local.aks_cluster_name
  depends_on = [
    azurerm_kubernetes_cluster.aks-cluster,
  ]
}
data "azuread_service_principal" "aks-enterprise-application-managed-id" {
  display_name = "${local.aks_cluster_name}-agentpool"
  depends_on = [
    azurerm_kubernetes_cluster.aks-cluster,
  ]
}
```
To:
```hcl
locals {
  cluster_principal_id   = azurerm_kubernetes_cluster.aks-cluster.identity[0].principal_id
  agentpool_principal_id = azurerm_kubernetes_cluster.aks-cluster.kubelet_identity[0].object_id
}

data "azuread_service_principal" "aks-enterprise-application" {
  object_id = local.cluster_principal_id      
}
data "azuread_service_principal" "aks-enterprise-application-managed-id" {
  object_id = local.agentpool_principal_id
}
```
the data sources depend on local variables instead, which will not trigger an updateof data/role_assignment when principal_id is same during the second apply, and trigger an update when principal_id changes

**Comment by michelefa1988** [comment]
Following since I have the same rather annoying situation in AKS

```
data "azurerm_kubernetes_cluster" "aks-read" {
  name                = var.name
  resource_group_name = var.resource_group

  depends_on = [
    azurerm_kubernetes_cluster.aks
  ]
}

resource "azurerm_role_assignment" "test" {
  role_definition_name = "AcrPull"
  principle_id         = data.azurerm_kubernetes_cluster.aks-read.kubelet_identity
  scope                = xxxx
}
```

If I update something minor in AKS (e.g update a firewall rule), then the role assignment gets recreated. 

If anyone knows any workarounds let me me know :)

**Comment by myc2h6o** [comment]
Hi @michelefa1988 from the config you provided, you are creating a `azurerm_role_assignment` with `kubelet_identity`. Instead of using an additional data source, I think you can try the above workaround which also did the similar thing

```hcl
locals {
  agentpool_principal_id = azurerm_kubernetes_cluster.aks-cluster.kubelet_identity[0].object_id
}

data "azuread_service_principal" "aks-enterprise-application-managed-id" {
  object_id = local.agentpool_principal_id
}

resource "azurerm_role_assignment" "test" {
  role_definition_name = "AcrPull"
  principal_id         = data.azuread_service_principal.aks-enterprise-application-managed-id.id
  scope                = xxxx
}
```

**Comment by michelefa1988** [comment]
HI @myc2h6o 

I have tried the code you send but it still didnt work.  Any other ideas? :(
```
# module.env-definition.module.identity_operator_nodepool_rbac.azurerm_role_assignment.role_assignment must be replaced
-/+ resource "azurerm_role_assignment" "role_assignment" {
      ~ id                               = "xxxx" -> (known after apply)
      ~ name                             = "xxx-xxx-xxx-xxx-xxx" -> (known after apply)
      ~ principal_id                     = "xxx-xxx-xxx-xxx-xxxxxx" -> (known after apply) # forces replacement
      ~ principal_type                   = "ServicePrincipal" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/xxxxx-xxxxx-xxxx-xxxxx-xxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxx-xxxxx-xxxxx-xxxxxx" -> (known after apply)
      ~ scope                            = "/subscriptions/xxxx/resourceGroups/MC_rg-xxxx-dev-weeu_aks-xxxx-xxx-weeu_westeurope" -> (known after apply) # forces replacement
      + skip_service_principal_aad_check = (known after apply)
        # (2 unchanged attributes hidden)
    }

```

**Comment by myc2h6o** [comment]
Hi @michelefa1988 I saw `scope` is also marked as `# forces replacement` in the diff, is this diff for the resource you are sharing? And if the property depends on a module output, this workaround shall not be needed.

**Comment by michelefa1988** [comment]
hi @myc2h6o 

Yes, I have 3x RBAC rules which are replaced and the screenshot is only of one of them ( to keep things clear).

I see what you mean the scope forces a replacement, however the scope of `identity_operator_nodepool_rbac` is the following (nothing complex)
` scope = "${data.azurerm_subscription.current.id}/resourceGroups/${module.resource_group.OUTPUT_resource_group_name}"
`

The resource group is derived from a datasource `data.azurerm_resource_group.resource_group.name`, and I am not making any updates to the resource group

**Comment by myc2h6o** [comment]
Hi @michelefa1988 I took another look and found below config in your previous comment https://github.com/hashicorp/terraform-provider-azurerm/issues/13152#issuecomment-1124595568
```hcl
data "azurerm_kubernetes_cluster" "aks-read" {
  name                = var.name
  resource_group_name = var.resource_group

  depends_on = [
    azurerm_kubernetes_cluster.aks
  ]
}
```
If you change the resource `azurerm_kubernetes_cluster.aks`, even if the `principal_id` doesn't change, the whole data source `aks-read` is recalculated at plan phase and value is retrieved again at apply phase, so `principal_id` is still `known after apply` for this data source, even if it is known at plan phase for the original resource.

If this is the case, I would suggest reference the property value from the resource directly, thus the `principal_id` could be known at plan phase from the resource.

**Comment by michelefa1988** [comment]
@myc2h6o  Thank you very much!! Your message has solved my issue. 
The `depends_on` was there since first I wanted to create the cluster then read it. afterwards. However this caused the problem you mentioned. 
A simple workaround was to make the aks-read depends on a sleep which depends on azurerm_aks

<img width="515" alt="image" src="https://user-images.githubusercontent.com/12183072/170859077-931fbfe2-76f9-41d2-8767-0f5ebcb67df5.png">

**Comment by myc2h6o** [comment]
Hi @michelefa1988 Your workaround can resolve this force-replacement issue, however it brings a new issue: when those identities do change on the aks resource, the data source will still get the old values when running `terraform apply`, because before the update, the values don't change. This causes the data source and the followed resources not being updated. When you run another `terraform apply` after the first update is finished, the data source will get the updated identities and the remaining resources will then be updated. This is actually more error prone. Another thing is that, if the data source is changed to depend on the `time_sleep.wait_aks`, it could fail in a Create scenario since it could read before the aks resource is created.

It shall be same to read the values from the aks resource directly compared to reading it from a separate data source. Is there any specific reasons not reading it from the resource itself in your config?

**Comment by michelefa1988** [comment]
Hi @myc2h6o , you are right but how often do the identities change? Because in my opinion it is unlikely they will sudden change and if they do, them I can simply reapply the rbac rules - doing my best to choose the best of two evils :)

I didn't understand your last question? Did you mean why does the `aks-read` reference a sleep instead of `azurerm_aks
? If so the reason is since if I make the change below, then each time I update an IP address then the RBAC roles and recreated - which is what I am trying to avoid :(

![image](https://user-images.githubusercontent.com/12183072/170976426-c7cc5ed8-371a-4bdb-93cf-4875dc125d46.png)

**Comment by myc2h6o** [comment]
Hi @michelefa1988 for the module output like `module.aks.OUTPUT_aks_kubelet_identity[0].object_id`, you can use the property value directly from the resource `azurerm_kubernetes_cluster.aks`, instead of creating another data source `aks-read`. This will fix the force-replacement issue as well as making the dependency correct.

**Comment by michelefa1988** [comment]
@myc2h6o thats what I would have done but as far as I see in https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster the `kubelet_identity` block  is not exported so you need to read it out a datasource

**Comment by myc2h6o** [comment]
@michelefa1988 I'm able to find the `kubelet_identity` [here](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#object_id) in the doc which has the `object_id`, and you can find the usage in the previous comment https://github.com/hashicorp/terraform-provider-azurerm/issues/13152#issuecomment-1050648704. This is being added to the doc by #12037 and released in [v2.63.0](https://github.com/hashicorp/terraform-provider-azurerm/releases/tag/v2.63.0). If this is any property in data source but missing in resource, we shall be able to add the support in resource as well.

**Comment by rcskosir** [comment]
Thanks for opening this issue. This was a problem in the 2.x version of the provider which is no longer actively maintained.
If this is still an issue with the 3.x version of the provider please do let us know by opening a new issue, thanks!
